### DISCUSSION TITLE
Fix for transitive dev dependencies in monorepo

### DISCUSSION THREAD

**Discussion by vvasin** [pnpm/pnpm#7452] [Q&A] [answered]
Recently I've faced a problem regarding dependencies with different versions.
I've created an example repo where we can reproduce it: https://github.com/vvasin/pnpm-transitive-dev-deps-issue

Let's assume we have a workspace with 2 projects:
- the project `a` depending on `axios@1` and `axios-retry@3`
- the project `b` depending on `axios@0`

The package `axios-retry@3` have a development dependency `axios@1`, but doesn't have it as a peer dependency. It uses the dependency for [type definitions](https://github.com/softonic/axios-retry/blob/v3/index.d.ts#L1). When we build the project `a`, TypeScript considers these types as well. The problem is that the types are imported from the hoisted version of `axios`.

In pnpm, depending on which package's deps was installed last, those deps are hosited. So if we install deps for the project `a` first, the `axios@1` package will be hoisted. Then, if we install deps for the project `b`, the `axios@0` package will be hoisted, replacing the previous one. After that, if we build the project `a`, we will get the following error:
```
src/index.ts:4:12 - error TS2345: Argument of type 'AxiosStatic' is not assignable to parameter of type 'AxiosInstance | AxiosStatic'.
```
It happens because TypeScript uses types from `axios@1` for the project and tries to match them with types imported from the hoisted `axios@0` by `axios-retry`.
By the way, when we install deps for both projects at once, pnpm hoists either `axios@1` or `axios@0` (sometimes).

I assume, there is a bug in the `axios-retry` package, and the best way is to rely on the fixed version with added `axios` as a peer dependency. By the way, the version 4 released lately already contains [this fix](https://github.com/softonic/axios-retry/blob/master/package.json#L35), but let's assume it was not released yet.

Is there a workaround to fix it on our side? Maybe it is possible to configure a project-level hoisting? Or force inject the peer dependency? Or make pnpm consider dev dependencies? I couldn't find the solution.

**Comment by await-ovo** [comment, marked as answer]
We can use `.pnpmfile.cjs` to mutate `axios-retry`'s  `package.json` like following:

```
function readPackage(pkg, context) {
  if (pkg.name === 'axios-retry') {
    pkg.peerDependencies = {
      ...pkg.peerDependencies,
      axios: '^1.2.3'
    }
  }
  return pkg
}
module.exports = {
  hooks: {
    readPackage
  }
}
```
